### PR TITLE
[SPARK-12420][SQL] Have a built-in CSV data source implementation

### PR DESCRIPTION
**Pull request by falaki**
CSV is the most common data format in the "small data" world. It is often the first format people want to try when they see Spark on a single node. Having to rely on a 3rd party component for this leads to poor user experience for new users. This PR merges the popular spark-csv data source package (https://github.com/databricks/spark-csv) with SparkSQL.

This is a first PR to bring the functionality to spark 2.0 master. We will complete items outlines in the design document (see JIRA attachment) in follow up pull requests.

Spark-csv was developed and maintained by several members of the open source community:
@dtpeacock: Type inference
@mohitjaggi: Integration with uniVocity-parsers 
@JoshRosen: Build and style checking 
@aley: Support for comments
@pashields: Support for compression codecs
@HyukjinKwon: Several bug fixes
@rbolkey: Tests and improvements
@huangjs: Updating API
@vlyubin: Support for insert
@brkyvz: Test refactoring
@rxin: Documentation
@andy327: Null values
@yhuai: Documentation
@akirakw: Documentation
@dennishuo: Documentation
@petro-rudenko: Increasing max characters per-column
@saurfang: Documentation
@kmader: Tests
@cvengros: Documentation
@MarkRijckenberg: Documentation
@msperlich: Improving compression codec handling
@thoralf-gutierrez: Documentation
@lebigot: Documentation
@sryza: Python documentation
@xguo27: Documentation
@darabos: License text in build file
@jamesblau: Nullable quote character
@karma243: Java documentation
@gasparms: Improving double and float type cast
@MarcinKosinski: R documentation
@addisonj: nullValue parsing

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+import scala.util.Try
+import scala.util.control.Exception._
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.types._
+
+private[sql] object CSVInferSchema {
+
+  /**
+    * Similar to the JSON schema inference. [[org.apache.spark.sql.json.InferSchema]]
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: Oh, actually I found several trivial typos. Here just a broken link here from `[[org.apache.spark.sql.json.InferSchema]]` to `[[org.apache.spark.sql.execution.datasources.json.InferSchema]]`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+    *     1. Infer type of each row
+    *     2. Merge row types to find common type
+    *     3. Replace any null types with string type
+    */
+  def apply(tokenRdd: RDD[Array[String]], header: Array[String]): StructType = {
+
+    val startType: Array[DataType] = Array.fill[DataType](header.length)(NullType)
+    val rootTypes: Array[DataType] = tokenRdd.aggregate(startType)(inferRowType, mergeRowTypes)
+
+    val stuctFields = header.zip(rootTypes).map { case (thisHeader, rootType) =>
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: And here `stuctFields` maybe for `structFields`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    val schemaFields = schema.fields
+    val requiredFields = StructType(requiredColumns.map(schema(_))).fields
+    val safeRequiredFields = if (params.dropMalformed) {
+      // If `dropMalformed` is enabled, then it needs to parse all the values
+      // so that we can decide which row is malformed.
+      requiredFields ++ schemaFields.filterNot(requiredFields.contains(_))
+    } else {
+      requiredFields
+    }
+    //println("safeRequiredFields: " + safeRequiredFields.mkString("\n\t"))
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: Lastly, here I believe this comment is meant to be removed.

**Comment by steveloughran** [comment]
Is this going to require the new parser JAR on the classpath everywhere, or will everything excluding CSV parsing still work without it?

**Comment by falaki** [comment]
@steveloughran this adds uniVocity-parsers as a dependency to SparkSQL. With assembly jar distribution, that jar will be included. If/When Spark moves away from assembly jar distribution, the uniVocity-parsers jar needs to be in classpath same way other dependencies should be.

**Comment by mohitjaggi** [comment]
this is great...thanks @falaki

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVParser.scala [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+          str.getChars((next - start).toInt, (next - start + n).toInt, cbuf, off)
+        }
+        next += n
+        if (n < len) {
+          val m = read(cbuf, off + n, len - n)  // have more space, fetch more input from iter
+          if(m != -1) n += m
+        }
+      }
+    }
+
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVParser.scala]: I think this is a trivial and ignorable stuff though.
If you are going to submit more commits, maybe this newline might better be removed.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+
+  test("save csv") {
+    withTempDir { dir =>
+      val csvDir = new File(dir, "csv").getCanonicalPath
+      val cars = sqlContext.read
+        .format("csv")
+        .option("header", "true")
+        .load(testFile(carsFile))
+
+      cars.repartition(1).write
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: A trivial and ignorable stuff though.
`coalesce(1)` maybe..?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+
+  test("save csv with quote") {
+    withTempDir { dir =>
+      val csvDir = new File(dir, "csv").getCanonicalPath
+      val cars = sqlContext.read
+        .format("csv")
+        .option("header", "true")
+        .load(testFile(carsFile))
+
+      cars.repartition(1).write
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: Ditto.

**Comment by HyukjinKwon** [comment]
Cool!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+   * Casts given string datum to specified type.
+   * Currently we do not support complex types (ArrayType, MapType, StructType).
+   *
+   * For string types, this is simply the datum. For other types.
+   * For other nullable types, this is null if the string datum is empty.
+   *
+   * @param datum string value
+   * @param castType SparkSQL type
+   */
+  private[csv] def castTo(
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: We should probably do this with expressions now that we are in Spark.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVParser.scala [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+      throw new IndexOutOfBoundsException()
+    } else if (len == 0) {
+      n = 0
+    } else {
+      if (next >= length) {   // end of input
+        n = -1
+      } else {
+        n = Math.min(length - next, len).toInt // lesser of amount of input available or buf size
+        if (n == length - next) {
+          str.getChars((next - start).toInt, (next - start + n - 1).toInt, cbuf, off)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVParser.scala]: does this actually do anything?

**Comment by falaki** [comment]
This PR was merged with https://github.com/apache/spark/pull/10766

**Pull request closed without merging**